tests/topotests/bgp_vpnv4_allocnh/test_bgp_vpnv4_allocnh.py

```diff
@@ -0,0 +1,386 @@
+                nexthopList.append(nh["interface"])
+            else:
+                assert (
+                    0
+                ), "{}, show mpls table, entry with neither nexthop nor interface detected".format(
+                    router.name
+                )
+
+
+def test_protocols_convergence():
+    """
+    Assert that all protocols have converged
+    statuses as they depend on it.
+    """
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    # Check BGP IPv4 routing tables on VRF1 of r1
+    logger.info("Checking BGP IPv4 routes for convergence on r1 VRF1")
+    router = tgen.gears["r1"]
+    json_file = "{}/{}/bgp_ipv4_routes_vrf1.json".format(CWD, router.name)
+    if not os.path.isfile(json_file):
+        assert 0, "bgp_ipv4_routes_vrf1.json file not found"
+
+    expected = json.loads(open(json_file).read())
+    test_func = partial(
+        topotest.router_json_cmp,
+        router,
+        "show bgp vrf vrf1 ipv4 json",
+        expected,
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=20, wait=0.5)
+    assertmsg = '"{}" JSON output mismatches'.format(router.name)
+    assert result is None, assertmsg
+
+    logger.info("Checking BGP VPNv4 routes for convergence on r2")
+    router = tgen.gears["r2"]
+    json_file = "{}/{}/bgp_vpnv4_routes.json".format(CWD, router.name)
+    if not os.path.isfile(json_file):
+        assert 0, "bgp_vpnv4_routes.json file not found"
+
+    expected = json.loads(open(json_file).read())
+    test_func = partial(
+        topotest.router_json_cmp,
+        router,
+        "show bgp ipv4 vpn json",
+        expected,
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=0.5)
+    assertmsg = '"{}" JSON output mismatches'.format(router.name)
+    assert result is None, assertmsg
+
+    # Check BGP labels received on r2
+    logger.info("Checking BGP VPNv4 labels on r2")
+    labelList = set()
+    bgp_vpnv4_table_check(tgen.gears["r2"], labelList)
+
+    # Check MPLS labels received on r1
+    mpls_table_check(tgen.gears["r1"], labelList)
+
+
+def test_flapping_bgp_vrf():
+    """
+    Turn down a remote BGP session
+    """
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+    logger.info("Unpeering BGP on r11")
+    dump = tgen.gears["r11"].vtysh_cmd(
+        "configure terminal\nrouter bgp 65500\nno neighbor 192.0.2.1\n",
+        isjson=False,
+    )
+
+    def _bgp_established(router):
+        output = json.loads(router.vtysh_cmd("show ip bgp vrf vrf1 neighbor 192.0.2.11 json"))
+        expected = {"192.0.2.11": {
+            "bgpState": "Established",
+            }
+        }
+        return topotest.json_cmp(output, expected)
+
+    def _bgp_failed(router):
+        output = json.loads(router.vtysh_cmd("show ip bgp vrf vrf1 neighbor 192.0.2.11 json"))
+        expected = {"192.0.2.11": {
+            "bgpState": "Active",
```

Comment:
actually, I think the first state is connect, then if there is ,no answer, the bgp session goes to active state.
the connect state does not last much time, so with the polling in progress, I am not sure it will match.

that is why I chose to use connect.